source.go

```diff
@@ -43,18 +56,215 @@ func (s *Source) Configure(ctx context.Context, cfg map[string]string) error {
 	return nil
 }
 
-func (s *Source) Open(ctx context.Context, pos sdk.Position) error {
+func (s *Source) Open(_ context.Context, pos sdk.Position) error {
+	s.client = notion.NewClient(notion.Token(s.config.token))
+	err := s.initPosition(pos)
+	if err != nil {
+		return fmt.Errorf("failed initializing position: %w", err)
+	}
+	return err
+}
+
+func (s *Source) initPosition(pos sdk.Position) error {
+	if len(pos) == 0 {
+		return nil
+	}
+	posParsed, err := time.Parse(string(pos), time.RFC3339)
+	if err != nil {
+		return fmt.Errorf("failed parsing time string %v: %w", string(pos), err)
+	}
+	s.lastEditedTime = posParsed
 	return nil
 }
 
 func (s *Source) Read(ctx context.Context) (sdk.Record, error) {
-	return sdk.Record{}, nil
+	err := s.populateIDs(ctx)
+	if err != nil {
+		return sdk.Record{}, fmt.Errorf("failed fetching page IDs: %w", err)
+	}
+	if len(s.fetchIDs) == 0 {
+		return sdk.Record{}, sdk.ErrBackoffRetry
+	}
+
+	return s.nextObject(ctx)
+}
+
+func (s *Source) nextObject(ctx context.Context) (sdk.Record, error) {
+	if len(s.fetchIDs) == 0 {
+		return sdk.Record{}, errors.New("no page IDs available")
+	}
+	pageID := s.fetchIDs[0]
+	s.fetchIDs = s.fetchIDs[1:]
+
+	// todo support databases
+	sdk.Logger(ctx).Debug().
+		Str("page_id", pageID).
+		Msg("fetching page")
+
+	block, err := s.client.Block.Get(ctx, notion.BlockID(pageID))
+	if err != nil {
+		return sdk.Record{}, fmt.Errorf("failed fetching page %v: %w", pageID, err)
+	}
+
+	children, err := s.getChildren(ctx, block)
+	if err != nil {
+		return sdk.Record{}, fmt.Errorf("failed fetching blocks for %v: %w", pageID, err)
+	}
+
+	record, err := s.blockToRecord(block, children)
+	if err != nil {
+		return sdk.Record{}, err
+	}
+	s.lastEditedTime = *block.GetLastEditedTime()
+	return record, nil
+}
+
+// getChildren gets all the child and grand-child blocks of the input block
+func (s *Source) getChildren(ctx context.Context, block notion.Block) ([]notion.Block, error) {
+	var children []notion.Block
+	if !block.GetHasChildren() {
+		sdk.Logger(ctx).Debug().
+			Str("block_id", block.GetID().String()).
+			Msg("block has no children")
+		return children, nil
+	}
+
+	fetch := true
+	var cursor notion.Cursor
+	for fetch {
+		resp, err := s.client.Block.GetChildren(
```

Comment:
I don't see any rate limiting anywhere ([info in docs](https://developers.notion.com/reference/request-limits#rate-limits)), does the client handle this behind the scenes? Otherwise a request might return an error and cause the pipeline to stop. I think this is quite important to get right if we want the connector to be stable.

Reply:
Lovro and I discussed this via Slack. While this is very important (any page which is slightly more complex will contain many blocks), we'll implement it in a follow-up PR (tracking issue #6), so that Samir is unblocked in his work.